### PR TITLE
get /tables endpoint

### PR DESCRIPTION
**Pull request by shermosa**

### CODE REVIEW

File: beekeeper-api/pom.xml [hunk 1]
```diff
@@ -20,6 +20,45 @@
...
+      <artifactId>specification-arg-resolver</artifactId>
+      <version>2.6.1</version>
+    </dependency>
+    <dependency>
+      <groupId>com.graphql-java</groupId>
+      <artifactId>graphql-spring-boot-starter</artifactId>
+      <version>5.0.2</version>
+    </dependency>
+    <dependency>
+      <groupId>com.expediagroup</groupId>
```
Comment (vedantchokshi) [inline beekeeper-api/pom.xml:28]: Can you put this dependency after the spring dependency? We tend to do our project ones at the top

File: beekeeper-api/pom.xml [hunk 2]
```diff
@@ -20,6 +20,45 @@
...
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.hamcrest</groupId>
+      <artifactId>hamcrest-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.springframework</groupId>
```
Comment (vedantchokshi) [inline beekeeper-api/pom.xml:53]: And spring ones at the top too

File: beekeeper-api/pom.xml [hunk 3]
```diff
@@ -20,6 +20,45 @@
...
+    <dependency>
+      <groupId>mysql</groupId>
+      <artifactId>mysql-connector-java</artifactId>
+      <version>8.0.18</version>
+    </dependency>
+    <dependency>
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
+    </dependency>
+    <dependency>
```
Comment (vedantchokshi) [inline beekeeper-api/pom.xml:44]: Can you add a blank line then comment saying test dependencies so the pom looks neat like the other modules?

File: beekeeper-api/pom.xml [hunk 4]
```diff
@@ -20,6 +20,45 @@
...
+      <groupId>org.hamcrest</groupId>
```
Comment (vedantchokshi) [inline beekeeper-api/pom.xml]: Can you remove this library and use assertThat from assertj like we are in the rest of the modules?

File: beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/HousekeepingEntityServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Service;
+
+import com.expediagroup.beekeeper.core.model.HousekeepingMetadata;
+import com.expediagroup.beekeeper.core.repository.HousekeepingMetadataRepository;
+
+@Service
+public class HousekeepingEntityServiceImpl implements HousekeepingEntityService<HousekeepingMetadata> {
```
Comment (vedantchokshi) [inline beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/HousekeepingEntityServiceImpl.java]: Maybe `HousekeepingMetadataService` sounds better?

File: beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java [hunk 6]
```diff
@@ -15,9 +15,23 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 
+import com.expediagroup.beekeeper.core.model.HousekeepingMetadata;
+
 @RestController
 public class BeekeeperController {
-  
+
+  private HousekeepingEntityService housekeepingEntityService;
```
Comment (vedantchokshi) [inline beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java]: Might be better to use the implementation class here because this service is specifically for `HousekeepingMetadata` so I don't think it should be generic.

File: beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java [hunk 7]
```diff
@@ -15,9 +15,23 @@
...
+
+  @RequestMapping("/tables")
```
Comment (vedantchokshi) [inline beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java]: Can you specify the type of mapping explicitly or change the annotation to `@GetMapping`? Would also be nice to explicitly state the arguments you're setting so its easy to know

File: beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/HousekeepingEntityService.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+
+import com.expediagroup.beekeeper.core.model.HousekeepingEntity;
+import com.expediagroup.beekeeper.core.model.HousekeepingMetadata;
+
+public interface HousekeepingEntityService<U extends HousekeepingEntity> {
+
+  Page<U> returnAllTables(Specification<HousekeepingMetadata> spec, Pageable pageable);
```
Comment (vedantchokshi) [inline beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/HousekeepingEntityService.java]: Can we call this `getAll` or `getAllEntities` as this is generic and not tied to `tables`?

File: beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java (hunk 7 above)
Comment (vedantchokshi) [inline beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java]: I think the endpoint `/housekeeping/tables` would be nicer. But ask Jose or other people for their opinion

File: beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java [hunk 9]
```diff
@@ -15,9 +15,23 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 
+import com.expediagroup.beekeeper.core.model.HousekeepingMetadata;
+
 @RestController
```
Comment (vedantchokshi) [inline beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java]: Below can we add `@RequestMapping("/api/v1")` so we can do versioning

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+  @Mock
+  private HousekeepingEntityService housekeepingEntityService;
+
+  private HousekeepingMetadata table1;
+  private HousekeepingMetadata table2;
+
+  @BeforeEach
+  public void createTables(){
+
+    LocalDateTime CREATION_TIMESTAMP = LocalDateTime.now(ZoneId.of("UTC"));
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java]: All caps means final static variable so it should be declared at the top, outside this method

Comment (vedantchokshi) [reply beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java]: If you're not using it any tests then you don't need to declare it

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+    table1 = new HousekeepingMetadata.Builder()
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java]: Can you make a utility/generator class with a static method which generates these objects for you? The method can only accept the stuff you want to set, the rest can be preset in the method because we don't care about it. Then you can just call the method and your code will look a lot neater.

Look at the repository tests if you want to see the example of the method

File: beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java [hunk 12]
```diff
@@ -15,9 +15,23 @@
...
 public class BeekeeperController {
-  
+
+  private HousekeepingEntityService housekeepingEntityService;
+
+  @RequestMapping("/tables")
+  public ResponseEntity<Page<HousekeepingMetadata>> getAllTables(Specification<HousekeepingMetadata> spec, Pageable pageable) {
```
Comment (vedantchokshi) [inline beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/BeekeeperController.java]: Can you also add the specification annotations properly so we have filtering?

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        .modifiedTimestamp(CREATION_TIMESTAMP)
+        .cleanupDelay(Duration.parse("P3D"))
+        .cleanupAttempts(0)
+        .lifecycleType(EXPIRED.toString())
+        .build();
+  }
+  
+  @Test
+  public void getAllTables() throws Exception {
+    List<HousekeepingMetadata> tablesList = new ArrayList<HousekeepingMetadata>();
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java]: We're using java 11 here so you can just do List.of(table1, table2) rather than adding them individually

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    System.out.println(tables.getContent());
+    when(housekeepingEntityService.returnAllTables(spec, pageable)).thenReturn(tables);
+
+    mockMvc
+        .perform(get(housekeepingEntityService.returnAllTables(spec, pageable).toString()))
+        .andDo(MockMvcResultHandlers.print())
+        .andExpect(status().isOk())
+        .andExpect(content().contentType(APPLICATION_JSON))
+        .andExpect(content().json(objectMapper.writeValueAsString(tables)));
+  }
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java]: Needs tests that check paging and sorting and also when returning empty

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    Page<HousekeepingMetadata> tables = new PageImpl<>(tablesList);
+    System.out.println(tables.getContent());
+    when(housekeepingEntityService.returnAllTables(spec, pageable)).thenReturn(tables);
+
+    mockMvc
+        .perform(get(housekeepingEntityService.returnAllTables(spec, pageable).toString()))
+        .andDo(MockMvcResultHandlers.print())
+        .andExpect(status().isOk())
+        .andExpect(content().contentType(APPLICATION_JSON))
+        .andExpect(content().json(objectMapper.writeValueAsString(tables)));
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java]: You should do verifications on the mock service to see that its only been called as expected

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+  @Mock
+  private Pageable pageable;
+
+  @BeforeEach
+  public void createTables(){
+    housekeepingEntityServiceImpl = new HousekeepingEntityServiceImpl(housekeepingMetadataRepository);
+
+    LocalDateTime CREATION_TIMESTAMP = LocalDateTime.now(ZoneId.of("UTC"));
+
+    table1 = new HousekeepingMetadata.Builder()
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java]: Use the utility class here now

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+
+import com.expediagroup.beekeeper.core.model.HousekeepingMetadata;
+import com.expediagroup.beekeeper.core.repository.HousekeepingMetadataRepository;
+
+@ExtendWith(MockitoExtension.class)
+public class HouseKeepingEntityServiceImplTest {
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java]: Rename this to the new name suggested

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java (hunk 11 above)
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java]: These also don't need to be generated in the beforeEach because all your tests won't require you to generate tables. Some will be testing errors and some will be returning empty lists. So its better to generate these tables inside your tests

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java (hunk 16 above)
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java]: These also don't need to be generated in the beforeEach because all your tests won't require you to generate tables. Some will be testing errors and some will be returning empty lists. So its better to generate these tables inside your tests

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        .modifiedTimestamp(CREATION_TIMESTAMP)
+        .cleanupDelay(Duration.parse("P3D"))
+        .cleanupAttempts(0)
+        .lifecycleType(EXPIRED.toString())
+        .build();
+  }
+
+  @Test
+  public void test(){
+
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java]: Remove blank line

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+        .creationTimestamp(CREATION_TIMESTAMP)
+        .modifiedTimestamp(CREATION_TIMESTAMP)
+        .cleanupDelay(Duration.parse("P3D"))
+        .cleanupAttempts(0)
+        .lifecycleType(EXPIRED.toString())
+        .build();
+  }
+
+  @Test
+  public void test(){
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java]: Your tests in this class might get long and they might test many cases so maybe pick a [good naming convention](https://dzone.com/articles/7-popular-unit-test-naming) from here and stick with that

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+        .creationTimestamp(CREATION_TIMESTAMP)
+        .modifiedTimestamp(CREATION_TIMESTAMP)
+        .cleanupDelay(Duration.parse("P3D"))
+        .cleanupAttempts(0)
+        .lifecycleType(EXPIRED.toString())
+        .build();
+  }
+  
+  @Test
+  public void getAllTables() throws Exception {
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/BeekeeperControllerTest.java]: Your tests in this class might get long and they might test many cases so maybe pick a [good naming convention](https://dzone.com/articles/7-popular-unit-test-naming) from here and stick with that

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+
+    List<HousekeepingMetadata> tables = new ArrayList<HousekeepingMetadata>();
+    tables.add(table1);
+    tables.add(table2);
+    when(housekeepingMetadataRepository.findAll(spec, pageable)).thenReturn(new PageImpl<>(tables));
+    Page<HousekeepingMetadata> result = housekeepingEntityServiceImpl.returnAllTables(spec,pageable);
+
+    assertThat(new PageImpl<>(tables),is(result));
+
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java]: Missing mockito verifications on your mocks

File: beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+  @Test
+  public void test(){
+
+    List<HousekeepingMetadata> tables = new ArrayList<HousekeepingMetadata>();
+    tables.add(table1);
+    tables.add(table2);
+    when(housekeepingMetadataRepository.findAll(spec, pageable)).thenReturn(new PageImpl<>(tables));
+    Page<HousekeepingMetadata> result = housekeepingEntityServiceImpl.returnAllTables(spec,pageable);
+
+    assertThat(new PageImpl<>(tables),is(result));
```
Comment (vedantchokshi) [inline beekeeper-api/src/test/java/com/expediagroup/beekeeper/api/HouseKeepingEntityServiceImplTest.java]: Change the assertThat to the one we use in the rest of the project

File: beekeeper-api/pom.xml [hunk 23]
```diff
@@ -20,6 +20,45 @@
...
+      <artifactId>beekeeper-core</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>mysql</groupId>
+      <artifactId>mysql-connector-java</artifactId>
+      <version>8.0.18</version>
+    </dependency>
+    <dependency>
+      <groupId>org.projectlombok</groupId>
```
Comment (vedantchokshi) [inline beekeeper-api/pom.xml]: You don't seem to be using lombok anywhere too

File: beekeeper-api/pom.xml [hunk 24]
```diff
@@ -20,6 +20,45 @@
...
+      <version>5.0.2</version>
+    </dependency>
+    <dependency>
+      <groupId>com.expediagroup</groupId>
+      <artifactId>beekeeper-core</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>mysql</groupId>
+      <artifactId>mysql-connector-java</artifactId>
```
Comment (vedantchokshi) [inline beekeeper-api/pom.xml]: Not sure but do you need this?

**Review by vedantchokshi — CHANGES_REQUESTED** [review]
Needs some changes

File: beekeeper-api/pom.xml [hunk 25]
```diff
@@ -20,6 +24,28 @@
...
+    </dependency>
+    <dependency>
+      <groupId>net.kaczmarzyk</groupId>
+      <artifactId>specification-arg-resolver</artifactId>
+      <version>${specification-arg-resolver.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.json</groupId>
+      <artifactId>json</artifactId>
+      <version>20080701</version>
```
Comment (vedantchokshi) [inline beekeeper-api/pom.xml:40]: Is this what resolved your JSON dependency issue?

File: beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/service/HousekeepingMetadataService.java [hunk 26]
```diff
@@ -13,11 +13,18 @@
...
-public class BeekeeperController {
-  
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+
+import com.expediagroup.beekeeper.core.model.HousekeepingEntity;
+import com.expediagroup.beekeeper.core.model.HousekeepingMetadata;
+
+public interface HousekeepingMetadataService<U extends HousekeepingEntity> {
```
Comment (vedantchokshi) [inline beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/service/HousekeepingMetadataService.java]: This should be called `HousekeepingEntityService`

File: beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/service/HousekeepingMetadataServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Service;
+
+import com.expediagroup.beekeeper.core.model.HousekeepingMetadata;
+import com.expediagroup.beekeeper.core.repository.HousekeepingMetadataRepository;
+
+@Service
+public class HousekeepingMetadataServiceImpl implements HousekeepingMetadataService<HousekeepingMetadata> {
```
Comment (vedantchokshi) [inline beekeeper-api/src/main/java/com/expediagroup/beekeeper/api/service/HousekeepingMetadataServiceImpl.java]: This should then be called `HousekeepingMetadataService`

File: beekeeper-api/pom.xml (hunk 25 above)
Comment (shermosa) [reply beekeeper-api/pom.xml:40]: yas queen
